### PR TITLE
Reporting PK and Unique constraints on complex datatypes in analyze and assessment

### PR DESCRIPTION
**Pull request by priyanshi-yb**
1. Reporting Indexes on the datatypes like `DATERANGE, TSRANGE, TSTZRANGE, NUMRANGE,INT4RANGE, INT8RANGE, INTERVAL (YEAR TO MONTH  and DAY TO SECOND) ` are not supported.
3. Reporting Unique and PRIMARY KEY constraints on all the unsupported index datatypes.
 https://yugabyte.atlassian.net/browse/DB-13973

jenkins -
https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/3990/
```
yugabyte=# create  table test_types(id int, t tsrange, d daterange, tz tstzrange, n numrange, i4 int4range, i8 int8range, s smallint, inym INTERVAL YEAR TO MONTH, inds INTERVAL DAY TO SECOND(6));
CREATE TABLE
yugabyte=# create index i1 on test_types(t);
ERROR:  INDEX on column of type 'user_defined_type' not yet supported
yugabyte=# create index i1 on test_types(d);
ERROR:  INDEX on column of type 'user_defined_type' not yet supported
yugabyte=# create index i1 on test_types(tz);
ERROR:  INDEX on column of type 'user_defined_type' not yet supported
yugabyte=# create index i1 on test_types(n);
ERROR:  INDEX on column of type 'user_defined_type' not yet supported
yugabyte=# create index i1 on test_types(i4);
ERROR:  INDEX on column of type 'INT4RANGE' not yet supported
yugabyte=# create index i1 on test_types(i8);
ERROR:  INDEX on column of type 'user_defined_type' not yet supported
yugabyte=# create index i1 on test_types (inym);
ERROR:  INDEX on column of type 'INTERVAL' not yet supported
yugabyte=# create index i1 on test_types (inds);
ERROR:  INDEX on column of type 'INTERVAL' not yet supported
yugabyte=# ALTER TABLE test_types ADD CONSTRAINT uk UNIQUE(d);
ERROR:  INDEX on column of type 'user_defined_type' not yet supported
yugabyte=# ALTER TABLE test_types ADD CONSTRAINT pk PRIMARY KEY(d);
NOTICE:  table rewrite may lead to inconsistencies
DETAIL:  Concurrent DMLs may not be reflected in the new table.
HINT:  See https://github.com/yugabyte/yugabyte-db/issues/19860. Set 'ysql_suppress_unsafe_alter_notice' yb-tserver gflag to true to suppress this notice.
ERROR:  PRIMARY KEY containing column of type 'user_defined_type' not yet supported

```

### CODE REVIEW

File: yb-voyager/cmd/analyzeSchema.go [hunk 1]
```diff
@@ -392,9 +392,11 @@ func checkStmtsUsingParser(sqlInfoArr []sqlInfo, fpath string, objType string) {
 			reportDeferrableConstraintCreateTable(createTableNode, sqlStmtInfo, fpath)
 			reportUnsupportedDatatypes(createTableNode.CreateStmt.Relation, createTableNode.CreateStmt.TableElts, sqlStmtInfo, fpath, objType)
 			parseColumnsWithUnsupportedIndexDatatypes(createTableNode)
+			reportUnsupportedConstraintsOnComplexDatatypesInCreate(createTableNode, sqlStmtInfo, fpath)
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:395]: Ideally, we should be adding code for new issues in queryissues 😬 
but this is urgent, so I think it's okay, we can take it up post release.

File: yb-voyager/cmd/analyzeSchema.go [hunk 2]
```diff
@@ -686,9 +688,97 @@ var UnsupportedIndexDatatypes = []string{
...
+	if !ok {
+		return
+	}
+	if alterCmd.GetSubtype() != pg_query.AlterTableType_AT_AddConstraint {
+		return
+	}
+	if !slices.Contains([]pg_query.ConstrType{pg_query.ConstrType_CONSTR_PRIMARY, pg_query.ConstrType_CONSTR_UNIQUE},
+		alterCmd.GetDef().GetConstraint().GetContype()) {
+		return
+	}
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:716]: nice early returns! easy to read!

File: yb-voyager/cmd/analyzeSchema.go [hunk 3]
```diff
@@ -686,9 +688,97 @@ var UnsupportedIndexDatatypes = []string{
...
+	columns := alterCmd.GetDef().GetConstraint().GetKeys()
+	for _, col := range columns {
+		colName := col.GetString_().Sval
+		typeName, ok := unsupportedColumnsForTable[colName]
+		if ok {
+			displayName := fmt.Sprintf("%s, constraint: %s", fullyQualifiedName, alterCmd.GetDef().GetConstraint().GetConname())
+			reportCase(fpath, fmt.Sprintf(ISSUE_CONSTRAINT_WITH_COMPLEX_DATATYPES, typeName), "https://github.com/yugabyte/yugabyte-db/issues/9698",
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:723]: the link only points to citext not being supported? Should we be creating a separate issue for this?

File: yb-voyager/cmd/analyzeSchema.go [hunk 4]
```diff
@@ -686,9 +688,97 @@ var UnsupportedIndexDatatypes = []string{
...
+	if !ok {
+		return
+	}
+	reportConstraintIfrequired := func(con *pg_query.Constraint, colNames []string, typeName string) {
+		conType := con.GetContype()
+		generatedConName := generateConstraintName(conType, fullyQualifiedName, colNames)
+		specifiedConstraintName := con.GetConname()
+		conName := lo.Ternary(specifiedConstraintName == "", generatedConName, specifiedConstraintName)
+		//report the PK / Unique constraint in CREATE TABLE on this column
+		if slices.Contains([]pg_query.ConstrType{pg_query.ConstrType_CONSTR_PRIMARY, pg_query.ConstrType_CONSTR_UNIQUE}, conType) {
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:746]: let's do an early return here as well? 
if not primary/unique constraint, return early.

File: yb-voyager/cmd/analyzeSchema.go [hunk 5]
```diff
@@ -686,9 +688,97 @@ var UnsupportedIndexDatatypes = []string{
...
+			summaryMap["TABLE"].invalidCount[fullyQualifiedName] = true
+			reportCase(fpath, fmt.Sprintf(ISSUE_CONSTRAINT_WITH_COMPLEX_DATATYPES, typeName), "https://github.com/yugabyte/yugabyte-db/issues/9698",
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:748]: same question about docs here

File: yb-voyager/cmd/analyzeSchema.go [hunk 6]
```diff
@@ -238,6 +237,7 @@ const (
 	POLICY_ROLE_ISSUE                              = "Policy require roles to be created."
 	VIEW_CHECK_OPTION_ISSUE                        = "Schema containing VIEW WITH CHECK OPTION is not supported yet."
 	ISSUE_INDEX_WITH_COMPLEX_DATATYPES             = `INDEX on column '%s' not yet supported`
+	ISSUE_CONSTRAINT_WITH_COMPLEX_DATATYPES        = `Primary key and Unique constraint on column '%s' not yet supported`
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:240]: ISSUE_PK_UK_CONSTRAINT_WITH_COMPLEX_DATATYPES

File: yb-voyager/cmd/analyzeSchema.go [hunk 7]
```diff
@@ -686,9 +688,97 @@ var UnsupportedIndexDatatypes = []string{
...
+		//report the PK / Unique constraint in CREATE TABLE on this column
+		if slices.Contains([]pg_query.ConstrType{pg_query.ConstrType_CONSTR_PRIMARY, pg_query.ConstrType_CONSTR_UNIQUE}, conType) {
+			summaryMap["TABLE"].invalidCount[fullyQualifiedName] = true
+			reportCase(fpath, fmt.Sprintf(ISSUE_CONSTRAINT_WITH_COMPLEX_DATATYPES, typeName), "https://github.com/yugabyte/yugabyte-db/issues/9698",
+				"Refer to the docs link for the workaround", "TABLE", fmt.Sprintf("%s, constraint: %s", fullyQualifiedName, conName), sqlStmtInfo.formattedStmt,
+				UNSUPPORTED_FEATURES, PK_UK_CONSTRAINT_ON_UNSUPPORTED_TYPE)
+		}
+	}
+	for _, column := range columns {
+		if column.GetColumnDef() != nil {
```
Comment (makalaaneesh) [inline yb-voyager/cmd/analyzeSchema.go:754]: just add comments in both the cases to explain again that in some cases constraint can be defined along with column, and in some cases, constraint is a defined separately.

File: yb-voyager/cmd/assessMigrationCommand.go [hunk 8]
```diff
@@ -940,13 +940,35 @@ func fetchUnsupportedPGFeaturesFromSchemaReport(schemaAnalysisReport utils.Schem
...
 	return unsupportedFeatures, nil
 }
 
+func getConstraintsOnComplexTypeUnsupportedFeature(schemaAnalysisReport utils.SchemaReport, unsupportedIndexDatatypes []string) UnsupportedFeature {
+	pkOrUkConstraintOnComplexTypesFeature := UnsupportedFeature{
+		FeatureName: "Primary / Unique key constraints on complex datatypes",
+		DisplayDDL:  false,
+		Objects:     []ObjectInfo{},
+	}
+	unsupportedIndexDatatypes = append(unsupportedIndexDatatypes, "array")             // adding it here only as we know issue form analyze will come with type
```
Comment (makalaaneesh) [inline yb-voyager/cmd/assessMigrationCommand.go:957]: Why can't we just get all issues in analyze-schema of type ISSUE_CONSTRAINT_WITH_COMPLEX_DATATYPES? Why do we have to filter here with unsupportedIndexDatatypes? that work was already done in analyze-schema, no?

File: yb-voyager/cmd/assessMigrationCommand.go [hunk 9]
```diff
@@ -940,13 +940,35 @@ func fetchUnsupportedPGFeaturesFromSchemaReport(schemaAnalysisReport utils.Schem
...
+	unsupportedIndexDatatypes = append(unsupportedIndexDatatypes, "user_defined_type") // adding it here as we UDTs will come with this type.
+	for _, unsupportedType := range unsupportedIndexDatatypes {
+		constraints := getUnsupportedFeaturesFromSchemaAnalysisReport(fmt.Sprintf("%s unique or pk constraints", unsupportedType), fmt.Sprintf(ISSUE_CONSTRAINT_WITH_COMPLEX_DATATYPES, unsupportedType),
```
Comment (makalaaneesh) [inline yb-voyager/cmd/assessMigrationCommand.go:960]: Remind me again why the reason is not constant and we add the data type to the reason? Just for display purposes?

File: yb-voyager/cmd/analyzeSchema.go (hunk 1 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go:395]: Yeah, so for this one I realized at the very last minute that we will have to new issues of constraints and I had to move all the index reporting functions there as I am maintaining state of `compositeTypes` and `unsupportedDataTypesColumns` map then I will have to move all of that which will take time. So thought will take up later.

File: yb-voyager/cmd/assessMigrationCommand.go (hunk 8 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/assessMigrationCommand.go:957]: Issue reason is format string where I will have to add type to get the issues

File: yb-voyager/cmd/assessMigrationCommand.go (hunk 9 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/assessMigrationCommand.go:960]: Yeah right

File: yb-voyager/cmd/analyzeSchema.go (hunk 3 above)
Comment (priyanshi-yb) [reply yb-voyager/cmd/analyzeSchema.go]: yeah created a new generic ticket and added that in all cases

File: yb-voyager/cmd/assessMigrationCommand.go [hunk 10]
```diff
@@ -940,35 +940,17 @@ func fetchUnsupportedPGFeaturesFromSchemaReport(schemaAnalysisReport utils.Schem
 	unsupportedFeatures = append(unsupportedFeatures, getUnsupportedFeaturesFromSchemaAnalysisReport(DEFERRABLE_CONSTRAINT_FEATURE, DEFERRABLE_CONSTRAINT_ISSUE, schemaAnalysisReport, false, ""))
 	unsupportedFeatures = append(unsupportedFeatures, getUnsupportedFeaturesFromSchemaAnalysisReport(VIEW_CHECK_FEATURE, VIEW_CHECK_OPTION_ISSUE, schemaAnalysisReport, false, ""))
 	unsupportedFeatures = append(unsupportedFeatures, getIndexesOnComplexTypeUnsupportedFeature(schemaAnalysisReport, UnsupportedIndexDatatypes))
-	unsupportedFeatures = append(unsupportedFeatures, getConstraintsOnComplexTypeUnsupportedFeature(schemaAnalysisReport, UnsupportedIndexDatatypes))
+
+	pkOrUkConstraintIssueSuffix := strings.Split(ISSUE_PK_UK_CONSTRAINT_WITH_COMPLEX_DATATYPES, "%s")[0]
```
Comment (makalaaneesh) [inline yb-voyager/cmd/assessMigrationCommand.go:944]: nit: substring/prefix instead of suffix

**Review by makalaaneesh — APPROVED** [review]
LGTM! with one comment

**Pull request merged**